### DISCUSSION TITLE
signIn() Redirects to an empty error when used

### DISCUSSION THREAD

**Discussion by Southclaws** [nextauthjs/next-auth#5336] [Help]
### Environment

  System:
    OS: macOS 12.5.1
    CPU: (8) arm64 Apple M1
    Memory: 62.84 MB / 16.00 GB
    Shell: 0.61.0 - /opt/homebrew/bin/nu
  Binaries:
    npm: 8.1.2 - ~/Library/Caches/fnm_multishells/12140_1662802635666/bin/npm
  Browsers:
    Chrome: 105.0.5195.102
    Firefox: 104.0.1
    Firefox Developer Edition: 100.0
    Safari: 15.6.1
  npmPackages:
    next: 12.3.0 => 12.3.0 
    next-auth: ^4.10.3 => 4.10.3 
    react: 18.2.0 => 18.2.0 

### Reproduction URL

private codebase

### Describe the issue

I just followed the docs and set up one simple /login page and I get an error with no actual error...

<img width="831" alt="image" src="https://user-images.githubusercontent.com/1636971/189480156-ada2ee98-b57c-4e93-a8c3-7482a1eaf0aa.png">

### How to reproduce

middleware is from docs:

```
export { default } from "next-auth/middleware";

export const config = { matcher: ["/", "/profile"] };
```

Auth config from docs with one minor change:

```
import NextAuth, { NextAuthOptions } from "next-auth";
import TwitterProvider from "next-auth/providers/twitter";

export const authOptions: NextAuthOptions = {
  providers: [
    TwitterProvider({
      clientId: process.env.TWITTER_ID,
      clientSecret: process.env.TWITTER_SECRET,
    }),
  ],
  pages: {
    signIn: "/login",
  },
  debug: true,
};

export default NextAuth(authOptions);
```

[index and other pages redirect to sign-in](https://next-auth.js.org/getting-started/client#redirects-to-sign-in-page-when-clicked) (couldn't figure out how to make this automatic for all pages except /login so currently I just have this copy pasted on every page lol)

```
import { useSession } from "next-auth/react";
import { signIn } from "next-auth/react";

export default function Page() {
  const { data: session, status } = useSession();

  if (status === "unauthenticated") {
    signIn();
  }

  return <div>{JSON.stringify(session)}</div>;
}
```

Is the latest version unstable? I just ran a `yarn upgrade` before but I set up this nextjs project today so I assume it would just be using all the latest stable versions. 

### Expected behavior

Expected it to work as the example, which I cloned to check and this works but can't seem to figure out what's different about my code copied from the docs.

**Comment by ThangHuuVu** [comment]
Hi, this sounds like a configuration error to me. Could you show some server logs? Help me to help you 🙌
For the middleware matcher, you need to include all pages you want to redirect if not logged in in the matcher array, and NextAuth.js will do redirection for you

**Reply by Southclaws** [reply]
I thought that's how it was meant to work, but adding these pages didn't result in redirects, that's why I'm calling signIn().

I'll see if I can rebuild a reproduction, I deleted the code for now as I just wanted to ship something so I built my own auth for that project now.

**Comment by Geczy** [comment]
i have the same issue and could not fix this after too much debugging

`api/auth/providers` just returns empty json, and then i get this error

here's the commit that started the issue https://github.com/dotabod/frontend/commit/1aaf25f00369a74071956316597964b14eacfede

if i rollback this commit, the issue goes away. also, the providers are empty randomly. i have to login, logout, repeatedly, to reproduce this problem. only when the provider json is empty do i get this error

**Reply by Geczy** [reply]
@ThangHuuVu perhaps this commit can help you see what the problem is?

**Reply by Geczy** [reply]
i undid everything in middleware and it still happens, so it must be my auth.ts file

**Reply by Geczy** [reply]
when i reset to 85bf2e3412fb4d6c315e3ef07bffed54cad56f00 it works 

so it could be any commit from 85bf2e3412fb4d6c315e3ef07bffed54cad56f00 to master in https://github.com/dotabod/frontend/commits/master

**Reply by Geczy** [reply]
https://github.com/dotabod/frontend/compare/85bf2e3412fb4d6c315e3ef07bffed54cad56f00..master here's the preview for what changed

**Reply by Geczy** [reply]
fixed after downgrading to 4 17

**Comment by joulev** [comment]
I faced this issue as well in v4.18.0. Downgrading to v4.17.0 fixed the issue, so I think there was some regression between v4.17 and v4.18.

**Reply by Geczy** [reply]
you're right, i downgraded to 4.17 and nothing is broken now lol

**Reply by austins** [reply]
Can confirm that this is an issue and reproducible with v4.18.0. Downgrading to v4.17.0 fixed it.

**Comment by GLips** [comment]
I had the same issue where `/api/auth/providers` would sometimes return blank. I found [a fix for using next-auth v4.18.0 here](https://github.com/t3-oss/create-t3-app/commit/c04a35a2b215c8537cf7f0d9ff9760c831720514).

Workaround seems to be related to `unstable_getServerSession`. Instead of
```
import { authOptions } from 'pages/api/auth/[...nextauth]';

unstable_getServerSession(
  ctx.req,
  ctx.res,
  authOptions
);
```

Create a new object with the `authOptions` values:
```
import { authOptions } from 'pages/api/auth/[...nextauth]';

unstable_getServerSession(
  ctx.req,
  ctx.res,
  { ...authOptions }
);
```

I'm not sure why spreading `authOptions` to a new object fixes the issue, but I'm no longer intermittently getting a blank result from the providers endpoint.

**Reply by GLips** [reply]
Possibly related to [this commit](https://github.com/nextauthjs/next-auth/commit/a307079e0fa4b4f1f1eae81140b8277773d4fa7e#diff-bdb80006ca8cdcac21aaed50b7882bbc7616cdcf5b693c9a40508dc3c06b4041R156)? I think @ammmze might be trying to fix it with [this one](https://github.com/nextauthjs/next-auth/commit/b19b2bcb3fa05614f3ac63b6828d310714834157).
